### DISCUSSION TITLE
ble.sh slow to start in recent versions

### DISCUSSION THREAD

**Discussion by Jai-JAP** [akinomyoga/ble.sh#524] [Q&A] [answered]
I am using `ble.sh` since a long time and recently the startup times are increasing, this is especially noticable in when using a debugger or some similar tools in VSCode. 

The commands sent to the terminal are executed after a while as in the below video.
[Recording 1](https://github.com/user-attachments/assets/77201222-8015-4a48-a66b-779671245cf3)
[Recording 1 with ble.sh disabled](https://github.com/user-attachments/assets/a4b51b0b-f87a-4bd8-8bba-4692635dd8b2)

This is especially troublesome when actually debugging / running a file as the first command is not run (venv doesn't get activated at all). This behaviour is shown below.

See that in the video below the venv doesn't get activated at all.
[Recording 2](https://github.com/user-attachments/assets/5b93a8e0-8932-47c5-85cd-a5a9fed8f28b)
[Recording 2 with ble.sh disabled](https://github.com/user-attachments/assets/df02baf2-c5c8-484a-bf67-2b96e167a0d8)

This is still the issue on a older version of vscode and extensions and it vanishes when disabling ble.sh. It was not an issue until recently and was working perfectly fine.

**Comment by akinomyoga** [comment]
> I am using `ble.sh` since a long time and recently the startup times are increasing,

Is that caused by changes in `ble.sh`? Or is the startup time increased by another factor? If it's the former, could you tell a commit without the issue?

> [Recording 2 with ble.sh disabled](https://github.com/user-attachments/assets/df02baf2-c5c8-484a-bf67-2b96e167a0d8).

As far as I see this, VS Code seems to send `^C` (<kbd>C-c</kbd>) after the `source` command, which cancels the execution of the command if the command hasn't been executed at the timing the shell receives `^C`. Currently, it just waits for a fixed time and sends `^C`, which I anticipate would cause a problem with any time-consuming `~/.bashrc`, and thus it is not a problem specific to ble.sh.

This seems like an issue with the VS Code (or the related plugin). It should activate the virtual environment through `--rcfile`, or it should properly sequence the timing of the execution of the previous command and `^C`.

**Reply by akinomyoga** [reply]
I haven't tried the actual behavior of VS Code, but I can reproduce the same problem when I input `echo hello[RET][C-c]` on the startup with `~/.bashrc` being `sleep 3` (without any other settings including ble.sh). It skips the execution of `echo hello`. The command `sleep 3` is just a command for the demonstration, but this means that any settings in `~/.bashrc` that takes time will cause the same issue.

@Jai-JAP Could you test if the same issue happens in the actual VS Code with the following single-line `~/.bashrc`?

```bash
# ~/.bashrc

sleep 1
```

**Reply by Jai-JAP** [reply]
I have tested with `sleep 3` in bashrc and it seems to work fine. the venv is indeed getting activated

[Recording 1](https://github.com/user-attachments/assets/a9b3575f-abaa-40f4-b30f-7720dd4e7cba)
[Recording 2](https://github.com/user-attachments/assets/750aba60-59d8-47e2-827e-43ee52bd1fe8)

@akinomyoga

**Reply by akinomyoga** [reply]
My laptop was broken in the middle of a business trip so I cannot do anything until I return to home on Nov 2nd night.

**Reply by Jai-JAP** [reply]
Any update on this? I am willing to help with this.

**Reply by akinomyoga** [reply]
I'm actually confused with your last result and unsure how to test it further. What I'm thinking now is that I'll probably need to set up a similar environment as yours and directly try to replicate the behavior in my environment.

**Reply by akinomyoga** [reply]
I have VS Code, but I'm not sure what to do next to get a similar setup as yours. I don't seem to have time to do guesswork about what you are actually using (e.g. for VS Code Extensions if any), what you've set up, and how to trigger it. It'd be helpful if you could explain to me how to replicate the environment with the problem in VS Code.

**Reply by Jai-JAP** [reply]
> I have VS Code, but I'm not sure what to do next to get a similar setup as yours. I don't seem to have time to do guesswork about what you are actually using (e.g. for VS Code Extensions if any), what you've set up, and how to trigger it. It'd be helpful if you could explain to me how to replicate the environment with the problem in VS Code.

I have the following VS Code setup for python apps.

Installed Extensions:
Python (`ms-python.python`)
Python Environment Manager (`donjayamanne.python-environment-manager`)


Create a new python file with `print("Hello world!")`


Now create the new venv by doing the following.
Open Command Pallete (<kbd>CTRL</kbd>+<kbd>SHIFT</kbd>+<kbd>P</kbd>), select `Python: Create Environment` and select env type as `Venv`. 

Now open the previously created python file and run it using python debugger. 
(Run icon on top right corner or run keybinding <kbd>CTRL</kbd>+<kbd>F5</kbd> in my case)

That should be enough to reproduce this situation.

**Reply by akinomyoga** [reply]
Thanks for the detailed setup. I could reproduce the problem. The problem arises with the following bashrc:

```bash
# bashrc
source /path/to/ble.sh
ble-attach
```

while it doesn't with the following bashrc:

```bash
# bashrc
source /path/to/ble.sh
```

This is unrelated to the speed of the initialization because I confirmed that the problem doesn't arise with the second setup no matter how slow the initialization is. Also, this is unrelated to the recent versions of ble.sh because I confirmed that the behavior is the same even in ble.sh of four years ago. So the initially provided information was mostly wrong guesses. Anyway, now that I could reproduce it in my environment, I can fix it on my side. Thanks.

**Reply by Jai-JAP** [reply]
Hey sorry for the late reply,

I am unsure about this as my setup doesn't have `ble-attach`. Maybe other parts of the setup includes it. 

Here are my config files.

.bashrc
```
# customisations

if [[ $- == *i* ]]; then
  bind 'set completion-ignore-case on'
  bind -x '"\C-l":clear'
  shopt -s autocd
  shopt -s cdspell
  shopt -s dirspell
  shopt -s expand_aliases
  set -C

  HISTCONTROL=ignoreboth

  if [[ "$TERM" != "linux" ]]; then
    _ls_icons="--icons"
  fi

  # shellcheck disable=SC2139
  alias ls="eza -ghoM --smart-group --git --no-time --no-permissions --group-directories-first --hyperlink --no-quotes -I .git ${_ls_icons}"
  # shellcheck disable=SC2139
  alias tree="eza -ghoMT --smart-group --git --no-time --no-permissions --group-directories-first --hyperlink --no-quotes -I .git ${_ls_icons}"
  alias cat="bat -p"

  help() {
    (command help "$@" 2>/dev/null || "$@" --help 2>/dev/null) | bat -pl help
  }

  if [[ $EUID == 0 ]]; then
    PS1='${SUDO_USER:+(\[\033[01;33m$SUDO_USER\033[0m\]) }\[\033[01;31m\]\u@\h\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\] \n \[\033[01;31m\]\$_\[\033[00m\] '
  else
    PS1='\[\033[01;32m\]\u@\h\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\] \n \[\033[01;32m\]\$_\[\033[00m\] '
  fi

  if ! shopt -q login_shell && [[ "$XDG_CURRENT_DESKTOP" == "GNOME" ]]; then
    alias logout="gnome-session-quit --no-prompt"
  fi

  if [[ "$PREFIX" =~ com.termux ]]; then
    # shellcheck disable=SC1091
    . "$HOME/.local/share/blesh/ble.sh"
  else # [[ "$TERM_PROGRAM" != "vscode" ]]; then
    # shellcheck disable=SC1091
    . "/usr/share/blesh/ble.sh"
  fi
fi

```

.blerc
```
#!/bin/bash

if [[ "$TERM_PROGRAM" != "vscode" && "$TERM" != "linux" ]]; then
  if [[ ! "$PREFIX" =~ com.termux ]]; then
    ble-import contrib/prompt-git
    #shellcheck disable=SC2016
    bleopt prompt_rps1='${_ble_contrib_prompt_git_base:+\q{contrib/git-branch} (\q{contrib/git-hash})}' \
      prompt_rps1_transient=1
    # RPS1='${_ble_contrib_prompt_git_base:+ \g{bold}| \g{reset}\q{contrib/git-branch} (\q{contrib/git-hash})}'

    if paru -Qq fzf &>/dev/null; then
      ble-import integration/fzf-completion
      ble-import integration/fzf-key-bindings

      _fzf_setup_completion path micro

      _ble_contrib_fzf_git_config=key-binding
      ble-import -d integration/fzf-git
    fi
  fi

  ble-bind -m emacs -f 'C-h' delete-backward-cword
  ble-bind -m emacs -f 'C-w' ''
elif [[ "$TERM" != "linux" ]]; then
  ble-bind -m emacs -f 'C-w' delete-backward-cword
  ble-bind -m emacs -f 'C-h' ''
else
  ble-bind -m emacs -f 'C-h' ''
  ble-bind -m emacs -f 'C-w' ''
  ble-bind -m emacs -f 'C-M-?' delete-backward-cword

  bleopt complete_auto_complete=
fi

# bleopt prompt_rps1="\$(RET=\$?; if [[ \$RET == 0 ]]; then echo \"\\g{bold,fg=green}\$RET\\g{reset}\"; else echo \"\\g{bold,fg=red}\$RET\\g{reset}\"; fi)$RPS1" \
#   prompt_rps1_transient=1

# ble-bind -f up 'history-search-backward immediate-accept'
# ble-bind -f down 'history-search-forward immediate-accept'

bleopt edit_vbell= \
  decode_error_char_vbell= \
  decode_error_cseq_vbell= \
  decode_error_kseq_vbell= \
  prompt_ruler=empty-line \
  tab_width=2 \
  edit_magic_accept=autocd \
  indent_offset=2 \
  indent_tabs=0

ble-face auto_complete='fg=240' \
  command_builtin='fg=86' \
  command_builtin_dot='fg=86' \
  command_function='fg=108,bold' \
  syntax_quotation='fg=124,bold' \
  syntax_quoted='fg=136,bold' \
  syntax_delimiter='fg=88,bold' \
  syntax_function_name='fg=108,bold' \
  region_match='fg=white,bg=160'

if [[ "$EUID" == 0 ]]; then
  FG_COLOR=160
else
  FG_COLOR=70
fi

if [[ "$TERM" != "linux" ]]; then
  _arrow="󰜴"
  _arrow_thin="➞ "
  _dir="󰉋"
else 
  _arrow="=>"
  _arrow_thin="->"
  _dir="\033[33;1mDIR changed\033[0m"
fi
bleopt prompt_ps1_final="\g{bold,fg=$FG_COLOR}${_arrow} \g{reset} "
blehook CHPWD="echo -ne '${_dir} ${_arrow_thin}\033[33;1m'; pwd | sed 's|'$HOME'|~|'; echo -ne '\033[0m'"

#shellcheck disable=SC2016
ble/function#advice around ble/widget/default/accept-line '
  ble-edit/content/clear-arg
  [[ ${_ble_edit_str//["$_ble_term_IFS"]} ]] && ble/function#advice/do'

#shellcheck disable=SC2016
ble/function#advice around ble/widget/discard-line '
  ble-edit/content/clear-arg
  [[ ${_ble_edit_str//["$_ble_term_IFS"]} ]] && ble/function#advice/do'
```

Let me know if you any more details for solving this issue.

**Reply by akinomyoga** [reply]
Thanks for the additional information, and I'm sorry for the delay in fixing it. I'm really busy recently.

Could you minimize your configuration while keeping it to reproduce the problem? Or all those settings are required to reproduce the problem?

**Reply by Jai-JAP** [reply]
I will share the minimal config required to reproduce this issue soon.

**Reply by Jai-JAP** [reply]
Hey I finally got some time after a busy week. 

I looked into minimizing the config while still reproducing the issue and looks like this is an issue even with an empty `.bashrc` (just sourcing ble.sh) and empty `.blerc`.

I am unsure why is there is a difference in the reproducibility of this issue. Possibly related to my system configuration which is as follows. (got it via <kbd>Ctrl</kbd>+<kbd>X</kbd>  <kbd>Ctrl</kbd>+<kbd>V</kbd>)

Empty configs (just sourcing ble.sh)
```
GNU bash, version 5.2.37(1)-release (x86_64-pc-linux-gnu) [Arch Linux]
ble.sh, version 0.4.0-devel4+31f264ad (noarch) [git 2.47.0, GNU Make 4.4.1, GNU Awk 5.3.1, API 4.0, PMA Avon 8-g1, (GNU MPFR 4.2.1, GNU MP 6.3.0)]
bash-completion, version 2.14.0 (hash:480ffcc6a751e55621ec526eb5dea7a0d86d9e72, 17877 bytes) (noarch)
locale: LANG=en_US.UTF-8 LC_ADDRESS=en_IN.utf-8 LC_IDENTIFICATION=en_IN.utf-8 LC_MEASUREMENT=en_IN.utf-8 LC_MONETARY=en_IN.utf-8 LC_NAME=en_IN.utf-8 LC_NUMERIC=en_IN.utf-8 LC_PAPER=en_IN.utf-8 LC_TELEPHONE=en_IN.utf-8 LC_TIME=en_IN.utf-8
terminal: TERM=xterm-256color wcwidth=12.1-west/16.0-2+ri, xterm:276 (0;276;0)
options: +extglob -hostcomplete +inherit_errexit
```

My setup
```
GNU bash, version 5.2.37(1)-release (x86_64-pc-linux-gnu) [Arch Linux]
ble.sh, version 0.4.0-devel4+31f264ad (noarch) [git 2.47.0, GNU Make 4.4.1, GNU Awk 5.3.1, API 4.0, PMA Avon 8-g1, (GNU MPFR 4.2.1, GNU MP 6.3.0)]
bash-completion, version 2.14.0 (hash:480ffcc6a751e55621ec526eb5dea7a0d86d9e72, 17877 bytes) (noarch)
locale: LANG=en_US.UTF-8 LC_ADDRESS=en_IN.utf-8 LC_IDENTIFICATION=en_IN.utf-8 LC_MEASUREMENT=en_IN.utf-8 LC_MONETARY=en_IN.utf-8 LC_NAME=en_IN.utf-8 LC_NUMERIC=en_IN.utf-8 LC_PAPER=en_IN.utf-8 LC_TELEPHONE=en_IN.utf-8 LC_TIME=en_IN.utf-8
terminal: TERM=xterm-256color wcwidth=12.1-west/16.0-2+ri, xterm:276 (0;276;0)
options: +noclobber +autocd +cdspell +dirspell +extglob +histappend -hostcomplete +inherit_errexit
```

EDIT: ble.sh was installed via the AUR package `blesh-git` and is up to date

**Reply by Jai-JAP** [reply]
Any progress on this? Let me know If you need any help from my side.

**Reply by akinomyoga** [reply]
I still cannot reproduce the problem.

One thing I noticed is that the command shown in your video, `source /home/user/test/.venv/bin/activate`, is not fetched in my environment at all even with the plain Bash without `ble.sh`. I confirmed it by directly modifying `.venv/bin/activate` to check whether `.venv/bin/activate` is loaded. The result is that `.venv/bin/activate` is not loaded at all. I tried three different ways to call the Python file, "Run Python File", "Run Python File in Dedicated Terminal", and "Python Debugger: Debug Python File", but none loaded `.venv/bin/activate`.

**edit**: However, it seems the necessary environment variables are already set up without calling `.venv/bin/activate` in my environment. Those environment variables seem to be initialized by `shellIntegration-bash.sh` through other VS Code environment variables.

**Reply by Jai-JAP** [reply]
That sounds weird, it should be activating the venv in both the cases. Have you selected the correct venv for the python interpreter. Check the bottom right corner.

![image](https://github.com/user-attachments/assets/b6ad26b0-224b-4b18-87a7-bd03db708381)

Also I have noticed one thing, this issue only occurs when running `Python Debugger: Debug Python File` Command. `Run Python File` seems unaffected from this issue. Possibly their launch script is doing something that breaks ble.sh

**Reply by akinomyoga** [reply]
It is activated for sure. On the bottom right corner, it says `3.10.6 ('.venv')`, and a tooltip pops up saying `./.venv/bin/python` when I hover a mouse cursor on `3.10.6 ('.venv')` on the bottom right corner.

**Without ble.sh**

![image](https://github.com/user-attachments/assets/0ed66d38-f3d8-44f9-aeca-ae34bb3ba316)

**With ble.sh**

![image](https://github.com/user-attachments/assets/fc8cfd44-880e-4c3f-bfd9-a1e062c42223)

**Reply by akinomyoga** [reply]
In both cases, the virtual environment is not set up for the launched shell session.

On the other hand, as far as I use "Run Python File" (instead of "Python Debugger: Debug Python File"), the virtual environment is set up both with and without `ble.sh`.

**Reply by akinomyoga** [reply]
- What is your  VS Code version?
- Also, what is your Python Extension version?

----

Here's mine.

```
Version: 1.96.2 (user setup)
Commit: fabdb6a30b49f79a7aba0f2ad9df9b399473380f
Date: 2024-12-19T10:22:47.216Z
Electron: 32.2.6
ElectronBuildId: 10629634
Chromium: 128.0.6613.186
Node.js: 20.18.1
V8: 12.8.374.38-electron.0
OS: Windows_NT x64 10.0.19045
```

![image](https://github.com/user-attachments/assets/d6d6ce6d-4c60-4709-8729-ba878775e3b3)

```
Identifier ms-python.python
Version 2024.22.1
Last Updated 2024-12-24, 16:44:34
Source VSIX
Size 47.5 MB
Cache 1.7 KB
```

**Reply by Jai-JAP** [reply, marked as answer]
Just did a bit more fiddling with the vscode configuration and it seems like the issue is indeed with the shell-integration breaking things. (They even mentioned that it breaks in complex situations)

![image](https://github.com/user-attachments/assets/5710dd4e-af32-4c7a-b4eb-add797d9c2cd)
https://code.visualstudio.com/docs/terminal/shell-integration#_automatic-script-injection

Now disabling the automatic loading of shell integrations and loading it manually seems to resolve the issue completely.

It is also the same if I completely disable the shell integration via the `terminal.integrated.shellIntegration.enabled` setting.

EDIT: Even after manually loading the shell-integration, it does not work at all with ble.sh. This is a completely different issue and VSCode cannot identify any thing within the terminal, not even commands, execution stauses, prompts etc.
At this moment, I would not bother waiting for this to fix as its barely a thing I use and went unnoticed since I started using ble.sh in VSCode. However it would be great to see ble.sh working flawlessly with VSCode and I am willing to help wherever possible in my available time.

Note: https://code.visualstudio.com/docs/terminal/shell-integration#_supported-escape-sequences

Please add to the docs/wiki mentioning that VSCode Integration is broken with ble.sh.

**Reply by Jai-JAP** [reply]
My VSCode version is
```
Version: 1.96.2
Commit: fabdb6a30b49f79a7aba0f2ad9df9b399473380f
Date: 2024-12-19T10:22:47.216Z
Electron: 32.2.6
ElectronBuildId: 10629634
Chromium: 128.0.6613.186
Node.js: 20.18.1
V8: 12.8.374.38-electron.0
OS: Linux x64 6.12.6-arch1-1
```

Python Extension version
```
Identifier: python.python
Version: 22.1
Last Updated: 12-24, 13:09:04
Size: 5 MB
Cache: 0 KB
```

**Reply by Jai-JAP** [reply]
I got some time to fiddle with the config and finally got the [manual loading of VSCode Shell Integration](https://code.visualstudio.com/docs/terminal/shell-integration#_manual-installation) to work. 

It needed adding the following lines to `.blerc`. (Transient Prompt enabled)
```
if [[ "$TERM_PROGRAM" == "vscode" ]]; then
  PS1_FINAL="\033]633;A\010\033[0m${PS1_FINAL}\033]633;B\010\033[0m"
  blehook PREEXEC+="echo -ne '\033]633;C\010\033]633;E\${1}\010\033[0m'; " \
    POSTEXEC+="echo -ne '\033]633;D;'\$?'\010\033[0m'; " \
    CHPWD+="echo -ne '\033]633;P;Cwd='\$(pwd)'\010\033[0m'; "
fi

bleopt prompt_ps1_final="$PS1_FINAL"
```

For Transient Prompt disabled instead, we need to add the following lines.
```
if [[ "$TERM_PROGRAM" == "vscode" ]]; then
  PS1="\033]633;A\010\033[0m${PS1}\033]633;B\010\033[0m"
  blehook PREEXEC+="echo -ne '\033]633;C\010\033]633;E\${1}\010\033[0m'; " \
    POSTEXEC+="echo -ne '\033]633;D;'\$?'\010\033[0m'; " \
    CHPWD+="echo -ne '\033]633;P;Cwd='\$(pwd)'\010\033[0m'; "
fi
```

**Comment by akinomyoga** [comment]
I took a look into the behavioral difference that I observed in https://github.com/akinomyoga/ble.sh/discussions/524#discussioncomment-11276124.

> Thanks for the detailed setup. I could reproduce the problem. The problem arises with the following bashrc:
> 
> ```shell
> # bashrc
> source /path/to/ble.sh
> ble-attach
> ```
> 
> while it doesn't with the following bashrc:
> 
> ```shell
> # bashrc
> source /path/to/ble.sh
> ```
> 
> This is unrelated to the speed of the initialization because I confirmed that the problem doesn't arise with the second setup no matter how slow the initialization is. Also, this is unrelated to the recent versions of ble.sh because I confirmed that the behavior is the same even in ble.sh of four years ago. So the initially provided information was mostly wrong guesses. Anyway, now that I could reproduce it in my environment, I can fix it on my side. Thanks.

It turned out to be caused by the VS shell integration. When one calls `ble-update` in the initialization file `~/.bashrc`, `ble.sh` requires it to be at the end of the initialization script. However, VS Code terminal seems to use a different initialization file at [`xxxx/vs/workbench/contrib/terminal/common/scripts/shellIntegration-bash.sh`](https://github.com/microsoft/vscode/blob/4fa5611d67dc84e105e9cd155a746f2d7813d9a0/src/vs/workbench/contrib/terminal/common/scripts/shellIntegration-bash.sh). Then, `~/.bashrc` is manually [sourced](https://github.com/microsoft/vscode/blob/4fa5611d67dc84e105e9cd155a746f2d7813d9a0/src/vs/workbench/contrib/terminal/common/scripts/shellIntegration-bash.sh#L17) from that `shellIntegration.sh`. Then, `shellIntegration-bash.sh` [modifies `PS1`](https://github.com/microsoft/vscode/blob/4fa5611d67dc84e105e9cd155a746f2d7813d9a0/src/vs/workbench/contrib/terminal/common/scripts/shellIntegration-bash.sh#L63) afterward. This breaks the requirement of `ble.sh`. Then, `PS1` that `ble.sh` recognizes is not updated by `shellIntegration-bash.sh`, which caused the difference I found above.

However, your issue seems to be unrelated to this one because you reported that you don't have `ble-attach` in your setting, though this problem shouldn't happen when you don't have `ble-attach`.
